ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/DiskSlots.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2020, Red Hat, Inc. and/or its affiliates,
+ * and individual contributors as indicated by the @author tags.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2020,
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com)
+ */
+package com.arjuna.ats.internal.arjuna.objectstore.slot;
+
+import com.arjuna.ats.arjuna.logging.tsLogger;
+
+import java.io.*;
+import java.util.zip.CRC32;
+
+/**
+ * File backed implementation of the SlotStore backend.
+ * WARNING: this is currently a test implementation and is NOT SAFE for production use yet.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-04
+ */
+public class DiskSlots implements BackingSlots {
+
+    private SlotStoreEnvironmentBean config;
+    private RandomAccessFile[] slots;
+
+    @Override
+    public synchronized void init(SlotStoreEnvironmentBean config) throws IOException {
+        if (this.config != null) {
+            throw new IllegalStateException("already initialized");
+        }
+
+        this.config = config;
+
+        File storeDir = new File(config.getStoreDir());
+        if (!storeDir.exists() && !storeDir.mkdirs()) {
+            throw new IOException(tsLogger.i18NLogger.get_dir_create_failed(storeDir.getCanonicalPath()));
+        }
+        slots = new RandomAccessFile[config.getNumberOfSlots()];
+        for (int i = 0; i < slots.length; i++) {
+            File file = new File(storeDir, "slot." + i);
+            slots[i] = new RandomAccessFile(file, "rw");
+        }
+    }
+
+    @Override
+    public void write(int slot, byte[] data, boolean sync) throws IOException {
+        RandomAccessFile randomAccessFile = slots[slot];
+        synchronized (randomAccessFile) {
+
+            randomAccessFile.seek(0);
+            randomAccessFile.writeInt(data.length);
```

Comment:
Right, the size check is in the common SlotStore code, not the backend. To make it apply to only some stores we'd have to move it into each relevant backend, but yes, that's possible.

Reply:
So we have a [bit of code](https://github.com/jbosstm/narayana/blob/master/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L875) that may serialize an XAResource into a slot. Will the default slot size be big enough for that or does the user need to be aware that the slot needs configuring properly?

Reply:
A typical JTA AtomicAction embedding 2 XAResource records and the key metadata weights in at under 1k. The slotSize defaults to 4k, which is a single block on most filesystems i.e. is the smallest size that the storage system handles anyhow. That should be enough headroom for most users. Records for e.g. XTS, JTS, subordinate tx and other more complex cases will vary more in size, but those should be caught early in staging environments. Bump the default to 8 or 16k if you're worried - for the current implementation there is no downsize to over-provisioning in that way.

Reply:
Well I was asking for input. If, as you say, it is typically 1K per resource then the default is sufficient. And if a user adds a resource that requires more then, since you throw an IOException if the slot is too small, it will be caught during testing and they can tweak the config appropriately. Thanks for the answer.